### PR TITLE
prevent endless async loop on Receive

### PR DESCRIPTION
**Pull request by dsyme**
@matthid I believe the fix https://github.com/Microsoft/visualfsharp/pull/3370 may have contained a mistake

I'm submitting what I believe to be the fix for a CI check. However I still need to understand why this wasn't caught by testing,

### CODE REVIEW

File: src/fsharp/FSharp.Core/control.fs [hunk 1]
```diff
@@ -2084,9 +2084,8 @@ namespace Microsoft.FSharp.Control
                             // If it isn't, then create it, and go back to the start to 
                             // check arrivals again.
                             match pulse with
-                            | null -> 
-                                if timeout >= 0 || cancellationSupported then 
-                                    ensurePulse() |> ignore
+                            | null when timeout >= 0 || cancellationSupported ->
+                                ensurePulse() |> ignore
                                 return! processFirstArrival()
```
Comment (matthid) [inline src/fsharp/FSharp.Core/control.fs]: I think the problem is that we need to call `ensurePulse` and than check arrivals a single time again before doing the actual wait (on the value returned by `ensurePulse`). While I think this patch is doing just that, it makes the code even more unreadable. But that might be some refactoring for another time.

**Comment by matthid** [comment]
Well, this code part is just written in a way too complicated way to ever get it bug-free (to be honest I need to look at this for quite some minutes every time again). But yeah, we probably miss some test-case here because this endless-loop wasn't noticed...

**Comment by dsyme** [comment]
@matthid I think it's actually quite hard to observe the endless async loop - in the normal case it just sits there requeueing receive items in the thread pool - burning some CPU but nothing else.  

The case where I noticed it was quite accidental - a mailbox processor loop accidentally switching to the UI thread, and then blocking the UI thread via this loop.  I've confirmed that this fix addresses that problem

I'm going to pull this once green

**Comment by matthid** [comment]
> a mailbox processor loop accidentally switching to the UI thread, and then blocking the UI thread via this loop. I've confirmed that this fix addresses that problem

Couldn't you do the same in a test-case?

**Comment by dsyme** [comment]
> Couldn't you do the same in a test-case?

Yes, I suppose so.

**Comment by matthid** [comment]
Offtopic:
So in the end this found another bug. Would be interesting to know how the other bug could have been prevented or found otherwise. It seems like `SwitchToContext` is quite unsafe to use and shouldn't be used without its counter-part? Do we need more "safe" versions of them?

**Comment by dsyme** [comment]
> Couldn't you do the same in a test-case?

Yes, thx

**Pull request merged**